3 - PB/Documentazione esterna/Specifica Tecnica/Specifica Tecnica.typ

```diff
@@ -146,7 +147,227 @@ Ciò avviene prestando attenzione alle tecnologie impiegate sia nel front-end ch
+
+=== Strategy
+Consente di definire una famiglia di algoritmi, incapsularli in classi separate e renderli intercambiabili. In questo modo è possibile applicare l'algoritmo appropriato senza dover conoscere i dettagli implementativi.
+
+
+
+== Classi e Componenti
+Ciascun layer possiede il suo indipendente sistema di classi e componenti e prevede metodi per comunicare con i layer adiacenti.
+
+=== Persistence layer
+Mediante delle Server Action offerte da Next.js, vengono eseguite delle query SQL atte alla lettura dei dati utili all'applicazione da un database esterno.
+Tali funzioni vengono utilizzate in:
+- `getAllBins`: ritorna le informazioni di tutti i bin lette dal database;
+- `getBinById`: dato un codice identificativo univoco, ritorna le informazioni relative al bin corrispondente lette dal database;
+- `getAllProducts`: ritorna le informazioni di tutti i prodotti lette dal database;
+- `getProductById`: dato un codice identificativo univoco, ritorna le informazioni relative al prodotto corrispondente lette dal database;
+- `SVGSanitize`: dato un path ad un file SVG caricato, ritorna il path del relativo file SVG sanificato, ovvero normalizzato e reso sicuro;
+- `readSavedSVG`: ritorna il contenuto del file SVG salvato su server;
+- `saveSVG`: dato il path di un file SVG, esso viene caricato sul server;
+- `getAllZones`: ritorna le informazioni di tutte le zone lette dal database;
+- `getBinsByZoneId`: dato un codice identificativo univoco, ritorna le informazioni relative a tutti i bin contenuti nella zona corrispondente lette dal database;
+- `getZoneById`: dato un codice identificativo univoco, ritorna le informazioni relative alla zona corrispondente lette dal database.
+
+
+Al fine di agevolare la divisione tra il Persistence layer ed il Business layer, viene applicato il pattern Repository mediante le classi:
+- l'interfaccia `DataRepositoryInterface`;
+- `binRepository`: implementa `DataRepositoryInterface` ed è responsabile dell'ottenimento dei dati relativi agli oggetti `Bin`;
+- `productRepository`: implementa `DataRepositoryInterface` ed è responsabile dell'ottenimento dei dati relativi agli oggetti `Product`;
+- `zoneRepository`: implementa `DataRepositoryInterface` ed è responsabile dell'ottenimento dei dati relativi agli oggetti `Zone`.
+
+Queste classi utilizzano le corrispondenti classi del pattern Data mapper:
+- l'interfaccia `DataMapperInterface`;
+- `binMapper`: implementa `DataMapperInterface` ed è responsabile della creazione di oggetti `Bin`;
+- `productMapper`: implementa `DataMapperInterface` ed è responsabile della creazione di oggetti `Product`;
+- `zoneMapper`: implementa `DataMapperInterface` ed è responsabile della creazione di oggetti `Zone`.
+
+
+#figure(
+  image("./imgs/Persistence Layer patterns class diagram.png", width: 100%),
```

Comment:
quello che lega ZoneMapper e BinMapper è un'aggregazione non una composizione. Infatti BinMapper è una classe che può vivere indipendentemente da ZoneMapper, tanto che è dichiarata e usata anche dentro a BinRepository. Proporrei discussione con gli altri a riguardo btw.

Reply:
per i diagrammi possiamo discuterne in meeting di retrospettiva